### DISCUSSION TITLE
Add healthcheck endpoint

### DISCUSSION THREAD

**Discussion by bkupidura** [home-assistant/architecture#650] [General]
Hello,
Currently HA is not exposing any endpoint which can be used to determine if HA is really working or not. It would be helpfull for any user deploying HA on top of k8s or docker. In case of failure, we can just simply restart container.

Health endpoint should be available without authentication.

How it can work:
* Every `N` seconds sends custom healthcheck event
* If we dont receive those events for `X` seconds, we should consider HA as unhealthy.

I already created PR for this (https://github.com/home-assistant/core/pull/58889).  As i was not aware about this repo :) Lets discuss if my approach is something which can be part of core.

**Comment by frenck** [comment]
I think the implementation in the upstream PR is not OK.

- Why does it need to configured?
- A health check should do something sane and actually check something. If you want to serve a static file, you can already do that. This logic doesn't need that.
- Home Assistant itself should not be responsible for considering itself unhealthy, that is up to an external process.
- It shouldn't be publicly accessible. Just because I run in a container, doesn't mean I want to publish the health check to the public together with the rest of my instance.

**Reply by bkupidura** [reply]
1) So i would assume that some people would like to disable healthcheck. So i would just stick to `healthcheck_enabled` and dont expose any other config options.
2) Problem with static_file is that there is no method to remove file in case of HA failure. In approach with dedicated endpoint, role of endpoint is just to calculate when last health check was observed. 
3) HomeAssistant will just provide information "i saw last event N seconds ago", so it will not do anything more. Its docker/k8s responsibility to handle this properly (e.g restart container)
4) Can you explain why this is a problem? Im not exposing any data except healthy/unhealthy. And most of servers are exposing health metrics without auth. Its not crucial, but exposing HA token to health monitoring would require some additional steps from users.

**Reply by frenck** [reply]
1. 👍 

> Problem with static_file is that there is no method to remove file in case of HA failure.

The current implementation checks nothing, thus at this point, a static file is exactly the same. Which is why 
I stated it should check something.

> HomeAssistant will just provide information "i saw last event N seconds ago",

Sure, but that is useless. Instead, check something to actually be able to state if Home Assistant is healthy.
Reason I aim for that, is because that would become a very interesting discussion. What defines if it is healthy? It responds to HTTP requests? What if databases are down? Recorder stopped working? Config errors? Maybe I really need my alarm to be available?

> Can you explain why this is a problem? Im not exposing any data except healthy/unhealthy.

A) It isn't needed for the outside world.
B) You can tell something about the version of Home Assistant if that is available (at least version X), meaning it leaks information.

**Reply by bkupidura** [reply]
> > Problem with static_file is that there is no method to remove file in case of HA failure.
> 
> The current implementation checks nothing, thus at this point, a static file is exactly the same. Which is why I stated it should check something.

Not really, static file is not suitable for health check. As HA would need to create and DELETE this file based on some checks. If something will break, HA will not delete this file so external tools are unable to determine if HA is working or not.

If we are talking about my PoC, HA will expose information 'i saw healtcheck event in last X seconds', and this can be iterpreted by external tools as:
1) Endpoint is not available at all - unhealthy
2) Endpoint is not responding for X seconds - unhealthy
3) Endpoint responded with 500 - unhealthy
4) Endpoint responded with 200 - healthy

Even if some HA internals will crash, it will be possible to discover this outside of HA.

I agree that checking just if message bus is working is not 100% accurate, but its better than nothing. During DST failure (https://github.com/home-assistant/core/issues/58783) my HA installation was not able to send events, so this healthcheck would detect that.

> > Can you explain why this is a problem? Im not exposing any data except healthy/unhealthy.
> 
> A) It isn't needed for the outside world. B) You can tell something about the version of Home Assistant if that is available (at least version X), meaning it leaks information.

OK, i dont agree with this. If somebody is exposing HA to internet without any WAF in front of it - he have bigger problem. But got yours point - we can add authentication in front of healthcheck endpoint.

**Reply by frenck** [reply]
> During DST failure (home-assistant/core#58783) my HA installation was not able to send events, so this healthcheck would detect that.

You reasoning is wrong, you don't need this health point to be able to detect that.

**Reply by frenck** [reply]
> OK, i dont agree with this.  If somebody is exposing HA to internet without any WAF in front of it - he have bigger problem

Peopel do, and we have a job to protect them the best we can. I will decline the PRs if that is not implemented. Health checks should not be available to the public.

**Reply by bkupidura** [reply]
Ok, so how we can detect if HA is running currently?

**Reply by frenck** [reply]
Define "HA is running"? That is the interesting part...

**Reply by bkupidura** [reply]
Core is running ;) For me would be:
* event_bus is working
* db connection is working
* http endpoint is available

And probably many, many other things im not aware

**Comment by thecode** [comment]
I think we should first start the discussion by defining what is considered an healthy Home Assistant. In my opinion knowing that http is accessible has very little benefit in defining if Home Assistant is healthy or not. 
Just two days ago the recorder stopped working for many of us and at least for me HTTP was working as usual, Do we consider this an healthy system?
If logfile is full with endless errors is it still considered healthy? If automation domain is broken?

As far as I understand it is already possible to query a certain endpoint without any implementation, so assuming you have an automation which just increment a counter and query the value externally, wouldn't it give a better result than a static HTTP health endpoint?

**Reply by bkupidura** [reply]
Please check my previous comment, during last outage - HA message bus was also not working. This implementation would detect that issue. Agree that its not perfect, but right now we dont have nothing.

**Reply by frenck** [reply]
>  This implementation would detect that issue.

That is actually not true, it would not have detected all issues for all people around this specific DST issue. Nevertheless, it would have detected yours. However, we are here to improve HA for everybody, not just your experience. This is what @thecode and me in the above as well, is commenting on.

We should explore and think about what a health check entails and how that should look for now and the future. Just implementing a minimal piece for your usecase isn't a proper way to approach this.

**Reply by bkupidura** [reply]
To be clear, im not saying that my PoC is checking every aspect of HA (as it is not, it check just very small piece of features). If anybody have better idea how to make this healthcheck endpoint - it would be great.

**Comment by allenporter** [comment]
My impression is that a health check endpoint in kubernetes is very very basic e.g. is the server ready to accept traffic.  Often examples can be simple 200 responses over http, or even tcp port health checks for non http servers.  See  [docs](https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes/).  We can do this part already.  I'm not sure that you want to have complex health checks at this level as this is used to restart and reschedule a job which doesn't help anything in a lot of cases.

The "interesting" parts are higher level.  I'd say the definition of healthy could exclude things like frontend and be limited to core, e.g. health at the API level.

IMO prometheus integration can use bearer tokens for auth and is a good approach.
E.g. 
[Servicemonitor](
https://github.com/k8s-at-home/charts/blob/adb11019c7555472ff240c2f2b7cda11ce5007d8/charts/stable/home-assistant/templates/servicemonitor.yaml)

A gap I have been thinking about  as an app level health check is related to integration health:

-  When I've previously had a messed up instance it appears like many of the integrations we're not able to load.
- This may be due to transient errors like a particular cloud integration is broken because they changed their API, or I screwed up recorder and many integrations are broken. 
- We had a discussion about add-on running/not running the other day and I asked about checking if integrations we're able to load, with this use case in mind.  I think of this as providing the same signal as the notification that says integrations failed to load.

On /how/ to go about this I'd say:
- container health checks exist already
- use prometheus for additional health checking
- as we define "what is healthy" ensure prometheus metrics cover it (maybe it's not the integration health thing I suggested but something else.

**Reply by bkupidura** [reply]
For sure prometheus can be used to detect different failures, but...
1) Not everybody use prometheus (i do)
2) Healthcheck by design are about service itself, Prometheus will be able to report that HA is broken, but thats it. It not be able to at least try to fix problem (restart container).

I would guess that for many people "self healing" is as important as detecting failure. If we will use externals tools to detect problem there will be no way to implement self healing.

**Comment by scop** [comment]
Food for thought re the format if/when the semantics of "what" have been worked out: RFC draft for HTTP health check endpoints: https://inadarei.github.io/rfc-healthcheck/

**Comment by allenporter** [comment]
If we're talking about k8s, my impression is there is a lot of prior art for how to do this.  It is pretty safe to align on prometheus as the defacto approach and we can make other integrations for other monitoring systems if needed.

Does restarting the container help with your recorder issue?  

Restarting home assistant would not help with the majority of issues I've personally run into (e.g. an intentional breaking change in a new build, I broke my config, a cloud provider took down their API). In each of these cases an alert would tell a human I need to take action such as rolling back the home assistant version (a rollback of a pr to a gitops repo, etc.)

I can imagine cases where you'd like home assistant to keep running when some integrations are broken and others are not broken, fwiw.

Aside: My previous way to tell that home assistant is down is not healthy is my lights don't turn on at sunset.

**Reply by bkupidura** [reply]
> If we're talking about k8s, my impression is there is a lot of prior art for how to do this. It is pretty safe to align on prometheus as the defacto approach and we can make other integrations for other monitoring systems if needed.

If we are thinking about notifing user that smth is broken, agree. If we think about self-healing prometheus would not help and we need to have health endpoint.

> 
> Does restarting the container help with your recorder issue?

Dont know as i didnt had healthcheck ;) But after OOMkill of HA container, it was back and alive. Problem was that OOM happend after 2h from initial symptoms. So not sure what would happend if HA would be restarted right after recorder failure.

> 
> Restarting home assistant would not help with the majority of issues I've personally run into (e.g. an intentional breaking change in a new build, I broke my config, a cloud provider took down their API). In each of these cases an alert would tell a human I need to take action such as rolling back the home assistant version (a rollback of a pr to a gitops repo, etc.)

At least couple of those examples will not happend in "production" environment. E.g wrong config, or broken build. Agree with external APIs.

> 
> I can imagine cases where you'd like home assistant to keep running when some integrations are broken and others are not broken, fwiw.

Agree. This is why in PR im checking most critical piece of HA (from my perspective). If message bus is not working, i would assume that nothing will work. Maybe im wrong.

> 
> Aside: My previous way to tell that home assistant is down is not healthy is my lights don't turn on at sunset.

**Comment by bkupidura** [comment]
So i double checked my monitoring, trying to find out if my approach would be able to detect last DST failure. And looks like answer is no, so you are right that checking only message bus is not enough to decide if HA is up or not. Its better than nothing, but its not sufficient. 

Some graphs:

Memory usage by HA POD:
<img width="1408" alt="Screenshot 2021-11-02 at 09 42 27" src="https://user-images.githubusercontent.com/4814684/139813702-40b5a3e4-6d25-46fb-9389-b0806d54a977.png">

Memory usage by Node-Red POD:
<img width="1413" alt="Screenshot 2021-11-02 at 09 43 08" src="https://user-images.githubusercontent.com/4814684/139813806-c80d02ec-f097-421a-897e-3af152f9ac81.png">


What happend is that after 2:00 AM (after 2:59->2:00 DST) HA start consuming all available resources, it was killed at ~4:00 by OOM.
After restart, everything was working as expected.

Currently my Node-Red is doing exactly same healthchecks against HA (fire event, receive event, if we miss N events - restart Node-Red trying to fix connectivity to HA). In graphs you can see that Node-Red detected problem with HA, but first detection happened around ~3:44. Which means that for ~1h 44m (2:00 till 3:44) HA was able to respond to message bus events, even when recorder (and possibly other componets) were broken. 
Probably failure of message bus was caused by resource limits for my HA installation (e.g memory limit). Without them, probably message bus would work as usual.

**Comment by frenck** [comment]
This architecture issue is old, stale, and possibly obsolete. Things changed a lot over the years. Additionally, we have been moving to discussions for these architectural discussions.

For that reason, I'm going to close this issue.

../Frenck
